<launch>

  <!-- args -->

  <arg name="INPUT_IMAGE" />

  <arg name="gui" default="true" />
  <arg name="GPU" default="0" />

  <!-- processing -->

  <node name="fcn_object_segmentation"
        pkg="grasp_prediction_arc2017" type="fcn_object_segmentation.py">
    <remap from="~input" to="$(arg INPUT_IMAGE)" />
    <rosparam subst_value="true">
      model_name: fcn32s
      n_class: 33
      model_file: $(optenv HOME)/data/mvtk/grasp_prediction_arc2017/logs/fcn32s_CFG-000_VCS-2400e9e_TIME-20170827-233211/models/FCN32s_iter00044000.npz
      class_agnostic: true
      gpu: $(arg GPU)
      thresh_class_prob: 0.5
      thresh_suction_prob: 0.7
    </rosparam>
  </node>

  <arg name="MANAGER" value="nodelet_manager" />
  <node name="$(arg MANAGER)"
        pkg="nodelet" type="nodelet" args="manager" />

  <!-- visualization -->

  <node name="label_image_decomposer_class"
        pkg="jsk_perception" type="label_image_decomposer.py">
    <remap from="~input" to="$(arg INPUT_IMAGE)" />
    <remap from="~input/label" to="/fcn_object_segmentation/output/label_class" />
    <rosparam>
      alpha: 0.5
      queue_size: 50
    </rosparam>
  </node>

  <node name="label_image_decomposer_suction"
        pkg="jsk_perception" type="label_image_decomposer.py">
    <remap from="~input" to="$(arg INPUT_IMAGE)" />
    <remap from="~input/label" to="/fcn_object_segmentation/output/label_suction" />
    <rosparam>
      alpha: 0.7
      queue_size: 50
    </rosparam>
  </node>

  <node name="image_view_prob_suction"
        pkg="image_view" type="image_view">
    <remap from="image" to="fcn_object_segmentation/output/prob_suction" />
    <remap from="~output" to="fcn_object_segmentation/output/prob_suction_view" />
    <rosparam>
      gui: $(arg gui) 
      min_image_value: 0
      max_image_value: 1
      colormap: 2  # JET
    </rosparam>
  </node>
  <node name="overlay_color_on_mono_prob_suction"
        pkg="nodelet" type="nodelet"
        args="load jsk_perception/OverlayImageColorOnMono $(arg MANAGER)">
    <remap from="~input/color" to="fcn_object_segmentation/output/prob_suction_view" />
    <remap from="~input" to="$(arg INPUT_IMAGE)" />
    <rosparam>
      color_alpha: 0.7
      queue_size: 50
    </rosparam>
  </node>

  <!-- gui -->

  <group if="$(arg gui)">
    <node name="image_view_rgb"
          pkg="image_view" type="image_view">
      <remap from="image" to="$(arg INPUT_IMAGE)" />
    </node>
    <node name="image_view_label_class"
          pkg="image_view" type="image_view">
      <remap from="image" to="label_image_decomposer_class/output/label_viz" />
    </node>
    <node name="image_view_label_suction"
          pkg="image_view" type="image_view">
      <remap from="image" to="label_image_decomposer_suction/output/label_viz" />
    </node>
  </group>

</launch>
